### PR TITLE
Use azd for deployment

### PR DESCRIPTION
**Pull request by pamelafox**
This pull request adds support for azd deployment, while still retaining the previous approach of just changing the .env file.

For the azd support, I added:
*  `infra` folder with Bicep files (mostly using AVM)
* `azure.yaml`
* `Dockerfile`
* `setup_intvect.py`: sets up a search index with integrated vectorization and uploads sample data to blob storage. Plus sh and ps1 files to call it.
* `write_env.sh` and `write_env.ps1`: Scripts to write the .env required for local running.

I also made some changes to existing files

* start.sh: I want to be able to call it from the root, without having to cd to app, so I made some changes to support that.
* [`app/backend/app.py`](diffhunk://#diff-5f1d71f916fd982ef5a7ab0b78e419d64d02a77c7422911db11911a74f3a8ee6R1-R61): Refactored to use `create_app` async function, added logging, and improved credential handling with `AzureDeveloperCliCredential` and `DefaultAzureCredential`.
* [`README.md`](diffhunk://#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5L8-R160): Extensive updates to include new sections for features, architecture diagram, deployment instructions, and cost and security guidance. This README now matches the requirements for our AI App Gallery templates.
* [`.vscode/settings.json`](diffhunk://#diff-a5de3e5871ffcc383a2294845bd3df25d3eeff6c29ad46e3a396577c413bf357L24-R25): Added `python.envFile` setting to the VS Code configuration, so that ".env"

### CODE REVIEW

**Comment by pamelafox** [comment]
I've marked this as draft so that I can go through the whole flow in Codespaces again.

**Comment by pamelafox** [comment]
I've gone through the flow in Codespaces. Unfortunately my swedencentral realtime deployments arent working today, despite eastus working. I *think* thats a service issue however.

File: infra/core/security/registry-access.bicep [hunk 1]
```diff
@@ -0,0 +1,19 @@
+metadata description = 'Assigns ACR Pull permissions to access an Azure Container Registry.'
```
Comment (jongio) [inline infra/core/security/registry-access.bicep:1]: Can you not use avm for this and role?

File: infra/main.parameters.json [hunk 2]
```diff
@@ -0,0 +1,78 @@
+{
```
Comment (jongio) [inline infra/main.parameters.json:1]: Can you use biccepparam file instead?

File: infra/core/host/container-app-upsert.bicep [hunk 3]
```diff
@@ -0,0 +1,130 @@
+metadata description = 'Creates or updates an existing Azure Container App.'
```
Comment (jongio) [inline infra/core/host/container-app-upsert.bicep:1]: FYI these will soon be pattern modules in AVM, prs are in progress, for now this is the best option.

File: infra/core/security/registry-access.bicep (hunk 1 above)
Comment (pamelafox) [reply infra/core/security/registry-access.bicep:1]: For role assignments, I wasn't sure which avm module was best. I found https://github.com/Azure/bicep-registry-modules/tree/main/avm/ptn/authorization/role-assignment but dont see examples of templates using it. Do you know the closest module?

File: infra/main.parameters.json (hunk 2 above)
Comment (pamelafox) [reply infra/main.parameters.json:1]: Good idea, porting!

File: infra/core/security/registry-access.bicep (hunk 1 above)
Comment (pamelafox) [reply infra/core/security/registry-access.bicep:1]: For registry access, given thats part of the container apps pattern, I'm inclined to leave that be.

File: infra/core/host/container-app-upsert.bicep (hunk 3 above)
Comment (pamelafox) [reply infra/core/host/container-app-upsert.bicep:1]: I saw those PRs, thanks for confirming.

File: infra/core/security/registry-access.bicep (hunk 1 above)
Comment (jongio) [reply infra/core/security/registry-access.bicep:1]: Okay

File: README.md [hunk 4]
```diff
@@ -5,118 +5,156 @@
...
+   * [Python >=3.11](https://www.python.org/downloads/)
+      * **Important**: Python and the pip package manager must be in the path in Windows for the setup scripts to work.
+      * **Important**: Ensure you can run `python --version` from console. On Ubuntu, you might need to run `sudo apt install python-is-python3` to link `python` to `python3`.
+   * [Git](https://git-scm.com/downloads)
+   * [Powershell](https://learn.microsoft.com/powershell/scripting/install/installing-powershell) - For Windows users only.
 
 2. Clone the repo (`git clone https://github.com/Azure-Samples/aisearch-openai-rag-audio`)
-4. The app needs to know which service endpoints to use for the Azure OpenAI and Azure AI Search. The following variables can be set as environment variables, or you can create a ".env" file in the "app/backend/" directory with this content.
-   ```
+3. Create a Python virtual environment and activate it.
```
Comment (chuwik) [inline README.md:57]: The 'start.sh/ps1' do this automatically now.

File: README.md [hunk 5]
```diff
@@ -5,118 +5,156 @@
...
+* Azure Container Apps: Consumption plan with 1 CPU core, 2.0 GB RAM. Pricing with Pay-as-You-Go. [Pricing](https://azure.microsoft.com/pricing/details/container-apps/)
+* Azure OpenAI: Standard tier, gpt-4o-realtime and text-embedding-3-large models. Pricing per 1K tokens used. [Pricing](https://azure.microsoft.com/pricing/details/cognitive-services/openai-service/)
+* Azure AI Search: Standard tier, 1 replica, free level of semantic search. Pricing per hour. [Pricing](https://azure.microsoft.com/pricing/details/search/)
+* Azure Blob Storage: Standard tier with ZRS (Zone-redundant storage). Pricing per storage and read operations. [Pricing](https://azure.microsoft.com/pricing/details/storage/blobs/)
+* Azure Monitor: Pay-as-you-go tier. Costs based on data ingested. [Pricing](https://azure.microsoft.com/pricing/details/monitor/)
+
+To reduce costs, you can switch to free SKUs for various services, but those SKUs have limitations.
+
+⚠️ To avoid unnecessary costs, remember to take down your app if it's no longer in use,
+either by deleting the resource group in the Portal or running `azd down`.
```
Comment (chuwik) [inline README.md:152]: nit: Azure Portal

File: app/backend/Dockerfile [hunk 6]
```diff
@@ -0,0 +1,11 @@
+FROM python:3.12-bullseye
```
Comment (chuwik) [inline app/backend/Dockerfile:1]: Should we use 3.11 to match the devcontainer? Or switch both to 3.12?

File: app/backend/setup_intvect.py [hunk 7]
```diff
@@ -0,0 +1,253 @@
...
+        logger.info(f"Index {index_name} already exists, not re-creating")
+    else:
+        logger.info(f"Creating index: {index_name}")
+        index_client.create_index(
+            SearchIndex(
+                name=index_name,
+                fields=[
+                    SearchableField(name="chunk_id", key=True, analyzer_name="keyword", sortable=True),
+                    SimpleField(name="parent_id", type=SearchFieldDataType.String, filterable=True),
+                    SearchableField(name="title", filterable=True),
```
Comment (chuwik) [inline app/backend/setup_intvect.py:89]: 'title' doesn't need to be filterable

File: app/backend/setup_intvect.py [hunk 8]
```diff
@@ -0,0 +1,253 @@
...
+            if filename in existing_blobs:
+                logger.info("Blob already exists, skipping file: %s", filename)
+            else:
+                logger.info("Uploading blob for file: %s", filename)
+                blob_client = container_client.upload_blob(filename, opened_file, overwrite=True)
+
+    # Start the indexer
+    try:
+        indexer_client.run_indexer(indexer_name)
+        logger.info("Indexer started. Any unindexed blobs should be indexed in a few minutes, check the Portal for status.")
```
Comment (chuwik) [inline app/backend/setup_intvect.py:213]: nit: Azure Portal

File: azure.yaml [hunk 9]
```diff
@@ -0,0 +1,37 @@
+# yaml-language-server: $schema=https://raw.githubusercontent.com/Azure/azure-dev/main/schemas/v1.0/azure.yaml.json
+
+name: azure-search-openai-demo
```
Comment (chuwik) [inline azure.yaml:3]: Should 'name' and 'template' be different, e.g. aisearch-openai-rag-audio? Not sure if this was copy/paste from other sample.

**Review by chuwik — COMMENTED** [review]
Left some comments.

File: app/backend/Dockerfile (hunk 6 above)
Comment (pamelafox) [reply app/backend/Dockerfile:1]: I'll change both to 3.12, given that 3.13 is now out (so 3.12 is considered fairly stable).

File: azure.yaml (hunk 9 above)
Comment (pamelafox) [reply azure.yaml:3]: Yes, thank you! Copypasta

File: README.md (hunk 4 above)
Comment (pamelafox) [reply README.md:57]: Good point!

**Review by chuwik — APPROVED** [review]

**Comment by pamelafox** [comment]
I'm doing another azd up test in a fresh Codespaces.

**Comment by pamelafox** [comment]
I had to revert the change from main.parameters.json to main.bicepparam as it was no longer popping up the OpenAI location picker. I then got a fully working app, both deployed and locally. This should be good to go!

**Pull request merged**